### DISCUSSION TITLE
Primeiras impressões sobre o NestJS

### DISCUSSION THREAD

**Discussion by synga** [frontendbr/forum#1473] [Perguntas e Respostas] [answered]
Fala galera, beleza?

To vendo que ultimamente estão falando muito sobre o [NestJS](https://docs.nestjs.com/) por ai, pessoas animadas e gostando do resultado.

Comecei agora a dar uma olhada na documentação, mas queria saber se alguém tem usado, se está gostando, se está sendo fácil de trabalhar, se pode falar da experiencia com ele, apontar pontos positivos e negativos.

Obrigado.

**Comment by lucashfreitas** [comment, marked as answer]
Opa!

Fala @synga , beleza?

O NestJS è um framework de backend que è utilizado principalmente pra construir web api's. Estou migrando um backend enorme pro NestJs. Estou tendo resultados maravilhosos com relação a produtividade e performance atè então. 

Ele foi desenvolvido inspirado no **Angular** e **Typescript**, então a sintaxe è bem parecida (apesar de ser utilizado pra construir backends). Ele utiliza o **express** por baixo dos panos mas tambèm è compativel com o fastify. 

**Sobre performance:**

O nest perde um pouquinhio comparado  ao express ou fastify puramente, segundo alguns benchmarks: [https://github.com/nestjs/nest/blob/master/benchmarks/all_output.txt](https://github.com/nestjs/nest/blob/master/benchmarks/all_output.txt)   [https://stackoverflow.com/questions/47733390/nestjs-vs-plain-express-performance](https://stackoverflow.com/questions/47733390/nestjs-vs-plain-express-performance). 

Segundo a essa pesquisa, o nestjs perde um pouco de performance comparado ao express puro, porem quando voce configura o nestjs pra usar o fastify, ele è ainda mais rapido que o express puro [https://docs.nestjs.com/techniques/performance](https://docs.nestjs.com/techniques/performance)

**Falando sobre produtividade:** O nestjs tem quase todos elementos necessarios para se construir uma API robusta jà prontos pra serem utilizados, como: 

- Suport a micro-servicos
- Aplicacao principal jà è integrada ao jest pra utilizar testes (porem vc pode utilizar qualquer lib que quiser pra fazer os testes tambem)
- Suporte a GraphQL
- Integracao com Swagger (Gera uma documentacao pra sua webapi com todas as rotas,etc)
- Autenticacao
- Validacao de Request
- Cache
- Cors,Csrf, etc (https://docs.nestjs.com/techniques/security)
- Camada de Repositorio (Utiliza o TypeORM ou Sequelize com suporte a MYSQL,MSSQL, MongoDB,etc) que tem uma documentacao incrivel. 
- Camada de Servicos, injecao de dependencia, etc , etc.
.
.
.
**-  E por ai vai......**

Lògico que quem vai determinar a arquitetura da sua API è o desenvolvedor. Jà trabalhei em projetos de Web Api's em Java, Go, PHP, .NET, NodeJS puro, etc mas o NestJs junto com o Typescript tem me surpreendido atè entao. 

Ele tem muitos recursos e componentes que è facil demais de adicionar a sua arquitetura e tambèm voce consegue fazer adaptacoes,  e melhor:

Ele è feito em Node, entao qualquer codigo em Node vai ser facilmente integrado <3

A minha impressao è que ele agiliza e jà tem muitos elementos prontos que sao necessarios pra construir uma web api robusta, alèm de ser desenvolvido em node, voce pode fazer adaptacoes tambem e etc.

**Outro ponto que ta fazendo muita diferença:**  (Nao è exclusivamente por causa do NestJs,  poderia ser utilizado com qualquer backend feito  em node): 

Tenho algumas aplicacoes frontend em Angular e React, e construi a minha api utilizando monorepository, e a minha camada de dominio que tem toda a descricao de servicos (interfaces), DTOs, interfaces, è utilizada tanto no back quanto to front nas minhas aplicacoes em React + Angular.

Lembrando que essa è uma visao e experiencia pessoal, e atè o momento nao vi nenhum ponto negativo relevante pra destacar, o desenvolvimento tà sendo bem produtivo, leve de rodar, facil de utilizar TDD, integracao continua, deploy continuo, etc.  

**Note:** Claro que a escolha de um framework/linguagem depende de vàrios fatores, como disponibilidade/habilidade da equipe & desenvolvedores, suporte, comunidade (_a do nest nao è enoooorme mas sempre que precisei fui respondido bem ràpido no git_) , documentacao (_achei a do nest bem eficiente_), etc.  Uma das vantagens que tambem notei è que meu front end em React , Angular utilizam a mesma linguagem que o meu backend (JS+TS) <3. Claro que tem muitos conceitos de backend que vao alèm da linguagem, etc, mas acredito que è uma vantagem de ter toda a stack utilizando uma linguagem sò :)

**Note:** O Nest ajuda, mas vai depender principalmente da equipe/devs pra construir uma boa arquitetura pra API, o que vai fazer MUITA diferenca em todo o andamento e futura manutençōes do projeto. 

Nao sei de empresas grandes utilizando, mas ja migrei grande parte do meu backend antigo pro NestJs e o resultado ta sendo incrível e acredito è um framework maduro/escalàvel pra ser adotado em producao (nao sei em relacao a larga, porque nao consegui achar nenhuma empresa realmente grande utilizando)

https://stackshare.io/nestjs
https://github.com/nestjs/nest/issues/1006

Espero que tenha ajudado! 

Cheers!

**Comment by synga** [comment]
@lucashfreitas Rapaz... você sanou todas as duvidas possíveis hahahaha

gostei dele pela semelhança ao Angular e pelo TS, batendo o olho no código do Hello World já consegui me situar, me pareceu bem fácil.
Nunca usei o Fastify, mas pelo que entendi só preciso é configurar pro Nest utilizar ele e tudo certo, né? Isso achei bem legal, não sei se mais pra frente vou precisar saber algo sobre o Fastify, mas de qualquer forma vou dar uma olhada melhor por que até então sempre usei o Express.
E como na empresa que trabalho sou só eu e mais 1 dev é algo legal pra se estudar e quem sabe utilizar mais pra frente ou migrar alguma API.

De qualquer forma muito obrigado, essa resposta foi de grande ajuda.

**Comment by lucashfreitas** [comment]
Show de bola @synga. =)

Sobre o fastfy è exatamente isso :)

Caso tenha qualquer dùvida em relação ao framework estou a disposição. Se você jà utiliza o angular no frontend, recomendo usar o Nest no backend!

As ordens!

**Comment by rbperegrino** [comment]
Uso o NestJS em projetos pessoais e onde trabalho estando migrando tudo que temos em C# para o NestJS com fastfy e typeorm, está tudo ocorrendo muito bem e muito rápido.

**Comment by oalexandrino** [comment]
top! muito rica essa discussão.

dúvidas pra galera:


Qual a  opinião para a seguinte Stack:

Angular + Nest.Js

Tenho pensado no backend de uma aplicação Angular e o NestJs é quase que igual em termos de sintaxe.

Indo mais longe e essa?

Angular + Nest.Js + TypeORM

Sobre o Express ou Fastify?

alguém poderia detalhar mais sobre eles?

**Comment by Coladela** [comment]
> Uso o NestJS em projetos pessoais e onde trabalho estando migrando tudo que temos em C# para o NestJS com fastfy e typeorm, está tudo ocorrendo muito bem e muito rápido.

Fiz o mesmo, de imediato tive mais aderência de outros devs que já estavam ambientados com js 
 e angular, nesse ponto foi muito positivo e rentável para o cliente, novos projetos ou uso node.js + ts, ou vou direto para o nest.

**Comment by Coladela** [comment]
> top! muito rica essa discussão.
> 
> dúvidas pra galera:
> 
> Qual a opinião para a seguinte Stack:
> 
> Angular + Nest.Js
> 
> Tenho pensado no backend de uma aplicação Angular e o NestJs é quase que igual em termos de sintaxe.
> 
> Indo mais longe e essa?
> 
> Angular + Nest.Js + TypeORM
> 
> Sobre o Express ou Fastify?
> 
> alguém poderia detalhar mais sobre eles?

Se for banco relacional, com certeza typeorm é muito produtivo, e já embarca padrões importantes de abstração, como "respositorio" por exemplo. Fiz um projeto que o client era ionic, com typeorm para o sqlite, para evolução de uma equipe de devs acho importante um framework que traga estruturação de código e padrões de arquitetura de uma forma natural, pois mesmo quando temos que fazer algo sem o nest ou ts, é natural reproduzir práticas que funcionam e escalam dentro de um consenso. meu desafio agora é implementar alguma coisa com docker e ter uma cobertura maior de testes.

**Comment by hugo-dutra** [comment]
> Opa!
> 
> Fala @synga , beleza?
> 
> O NestJS è um framework de backend que è utilizado principalmente pra construir web api's. Estou migrando um backend enorme pro NestJs. Estou tendo resultados maravilhosos com relação a produtividade e performance atè então.
> 
> Ele foi desenvolvido inspirado no **Angular** e **Typescript**, então a sintaxe è bem parecida (apesar de ser utilizado pra construir backends). Ele utiliza o **express** por baixo dos panos mas tambèm è compativel com o fastify.
> 
> **Sobre performance:**
> 
> O nest perde um pouquinhio comparado ao express ou fastify puramente, segundo alguns benchmarks: https://github.com/nestjs/nest/blob/master/benchmarks/all_output.txt https://stackoverflow.com/questions/47733390/nestjs-vs-plain-express-performance.
> 
> Segundo a essa pesquisa, o nestjs perde um pouco de performance comparado ao express puro, porem quando voce configura o nestjs pra usar o fastify, ele è ainda mais rapido que o express puro https://docs.nestjs.com/techniques/performance
> 
> **Falando sobre produtividade:** O nestjs tem quase todos elementos necessarios para se construir uma API robusta jà prontos pra serem utilizados, como:
> 
> * Suport a micro-servicos
> * Aplicacao principal jà è integrada ao jest pra utilizar testes (porem vc pode utilizar qualquer lib que quiser pra fazer os testes tambem)
> * Suporte a GraphQL
> * Integracao com Swagger (Gera uma documentacao pra sua webapi com todas as rotas,etc)
> * Autenticacao
> * Validacao de Request
> * Cache
> * Cors,Csrf, etc (https://docs.nestjs.com/techniques/security)
> * Camada de Repositorio (Utiliza o TypeORM ou Sequelize com suporte a MYSQL,MSSQL, MongoDB,etc) que tem uma documentacao incrivel.
> * Camada de Servicos, injecao de dependencia, etc , etc.
>   .
>   .
>   .
>   **-  E por ai vai......**
> 
> Lògico que quem vai determinar a arquitetura da sua API è o desenvolvedor. Jà trabalhei em projetos de Web Api's em Java, Go, PHP, .NET, NodeJS puro, etc mas o NestJs junto com o Typescript tem me surpreendido atè entao.
> 
> Ele tem muitos recursos e componentes que è facil demais de adicionar a sua arquitetura e tambèm voce consegue fazer adaptacoes, e melhor:
> 
> Ele è feito em Node, entao qualquer codigo em Node vai ser facilmente integrado <3
> 
> A minha impressao è que ele agiliza e jà tem muitos elementos prontos que sao necessarios pra construir uma web api robusta, alèm de ser desenvolvido em node, voce pode fazer adaptacoes tambem e etc.
> 
> **Outro ponto que ta fazendo muita diferença:** (Nao è exclusivamente por causa do NestJs, poderia ser utilizado com qualquer backend feito em node):
> 
> Tenho algumas aplicacoes frontend em Angular e React, e construi a minha api utilizando monorepository, e a minha camada de dominio que tem toda a descricao de servicos (interfaces), DTOs, interfaces, è utilizada tanto no back quanto to front nas minhas aplicacoes em React + Angular.
> 
> Lembrando que essa è uma visao e experiencia pessoal, e atè o momento nao vi nenhum ponto negativo relevante pra destacar, o desenvolvimento tà sendo bem produtivo, leve de rodar, facil de utilizar TDD, integracao continua, deploy continuo, etc.
> 
> **Note:** Claro que a escolha de um framework/linguagem depende de vàrios fatores, como disponibilidade/habilidade da equipe & desenvolvedores, suporte, comunidade (_a do nest nao è enoooorme mas sempre que precisei fui respondido bem ràpido no git_) , documentacao (_achei a do nest bem eficiente_), etc. Uma das vantagens que tambem notei è que meu front end em React , Angular utilizam a mesma linguagem que o meu backend (JS+TS) <3. Claro que tem muitos conceitos de backend que vao alèm da linguagem, etc, mas acredito que è uma vantagem de ter toda a stack utilizando uma linguagem sò :)
> 
> **Note:** O Nest ajuda, mas vai depender principalmente da equipe/devs pra construir uma boa arquitetura pra API, o que vai fazer MUITA diferenca em todo o andamento e futura manutençōes do projeto.
> 
> Nao sei de empresas grandes utilizando, mas ja migrei grande parte do meu backend antigo pro NestJs e o resultado ta sendo incrível e acredito è um framework maduro/escalàvel pra ser adotado em producao (nao sei em relacao a larga, porque nao consegui achar nenhuma empresa realmente grande utilizando)
> 
> https://stackshare.io/nestjs
> [nestjs/nest#1006](https://github.com/nestjs/nest/issues/1006)
> 
> Espero que tenha ajudado!
> 
> Cheers!

Estou usando. Uso Angular no Front-end e usava Laravel. Estou migrando o back-endo pro NestJS e esse framwork é foda bagarai....

**Comment by Pacheco95** [comment]
Eu comecei meu desenvolvimento no node com o express, depois migrei pro Adonis (fui direto pro 5) após ouvir alguns podcasts da Rocketseat e assistir uns vídeos no YT. Achei muito bom ter tarefas corriqueiras todas automatizadas na palma da mão. Porém, me apaixonei mesmo foi com o Nest. A forma como o nest organiza o código em módulos/features é muito melhor que no adonis, o qual já tem uma estrutura fixa que vc tem que seguir. Recentemente abri até uma [discussão](https://github.com/adonisjs/core/discussions/1940) no adonis perguntando se haveria uma forma de organizar a estrutura de pastas e arquivos em features como é feito no Nest.

Até agora, o que posso apontar de prós e contras de cada um é:

# Adonis 5
## prós
- CLI: o CLI do adonis, o `Ace`, é muito útil para criar controllers, providers, validators, etc.
- Comunidade: bem aquecida. Têm bastante dev no SO, GitHub e no Discord para sanar dúvidas
- Configurável: no adonis, quase tudo é configurável por meio de um provider. Se vc precisa configurar o CORS, as váriáveis de ambiente, os bancos de dados, ..., enfim, pra tudo que precisar de configuração é só criar um provider e tá tudo certo.

## contras
- Mantido por um exército de um homem só
- Organização: no adonis tudo já tem seu lugar. Para uns isso é o céu. Também achei que fosse, até meu projeto crescer e eu ficar perdido pra juntar os teste, controllers, provides, schemas, migrations, interfaces, ... de cada feature, cada um em um canto do projeto.
- A documentação ainda está precária. Tem poucos exemplos e/ou referências externas. Muitas vezes é necessário acessar a documentação da v4 para poder ter uma noção de como fazer determinada coisa.
- Dependências: o adonis têm muitas dependências internas que foram desenvolvidas pelo mesmo desenvolvedor do core, o Virk. Alguns módulos poderiam ter sido terceirizados para outras libs open source já existentes (como a lib de testes, por exemplo). Isso deixa a manutenção do core mais lenta e menos suscetível à PRs da comunidade.
- É production ready mas ainda é preview: nem todos os pacotes do adonis 4 foram migrados para o 5.

# Nest
## prós
- Dependency Injection: cara, eu me apaixonei com essa feature do nest. Eu já havia trabalhado com o Spring Boot no Java e já havia me acostumado com a forma de programar com DI. Ela torna a programação muito mais fluida e diminui muito o acoplamento de código além da facilidade de testar.
- CLI: não é tão boa quanto a do Adonis, mas também é muito útil e automatiza muitas tarefas corriqueiras. No entando, senti 
 muita falta de um comando para listar as rotas da aplicação igual ao `node ace list:routes` do adonis :heart: 
- Modularização: em conjunto com a DI, essa feature me conquistou. A forma como o nest lida com módulos é muito interessante. Ele consegue injetar as dependências de controllers, services, schemas, etc., tudo dentro de um módulo e ainda exportar todos esses recursos para que outros módulos possam usar. E com ela você organiza seus arquivos por feature ao invés de categoria.
- Testes: os testes também são modularizados. Basicamente os testes sobem uma versão do app com os módulos que você quer e vc n precisa subir uma aplicação inteira para poder testar uma simples feature.
- Decorators: outra feature muito bacana do nest. Pra você criar um controller basta criar uma classe e decorar ela com um `@Controller`, nada de ficar criando um arquivo `routes.ts` e abarrotar ele de rotas. Existem dezenas de decorators para se criar pipes de validação, autenticação, transforms, interceptors e o que mais vc quiser, inclusive criar seus proṕrios.

## contras
- Feedbacks de erros: isso não é muito legal no nest. Quando sua aplicação começa a crescer e você começa a ter dezenas de módulos conversando uns com os outros, vão aparecendo problemas de injeções de dependências que são difíiiiiceis de serem resolvidos só com as mensagens de erro que o nest te oferece. Ele te dá um norte e vc se vira com seus breakpoints pra achar a causa do erro (que geralmente é um export ou import num módulo ou um `@Inject` faltando).
- Listagem de rotas: como no nest não existe um `routes.ts`, fica difícil ter a visão de todas as rotas que você já tem na sua aplicação. Nisso o Adonis mata de letra.

Resumindo. São ótimos frameworks que batem de frente. Mas na minha opinião, se eu fosse iniciar um novo projeto hoje, eu ficaria com o Nest. Espero ter esclarecido algumas dúvidas  que a galera geralmente tem. Críticas e sugestões são muito bem vindas!
